controller/setting_controller.go

```diff
@@ -319,12 +313,35 @@ func (sc *SettingController) syncDangerZoneSettingsForManagedComponents(settingN
 	dangerSettingsRequiringSpecificDataEngineVolumesDetached := []types.SettingName{
 		types.SettingNameV1DataEngine,
 		types.SettingNameV2DataEngine,
+		types.SettingNameGuaranteedInstanceManagerCPU,
+		types.SettingNameV2DataEngineGuaranteedInstanceManagerCPU,
 	}
 
 	if slices.Contains(dangerSettingsRequiringSpecificDataEngineVolumesDetached, settingName) {
-		if err := sc.updateDataEngine(settingName); err != nil {
-			return errors.Wrapf(err, "failed to apply %v setting to Longhorn instance managers when there are attached volumes. "+
-				"It will be eventually applied", settingName)
+		switch settingName {
+		case types.SettingNameV1DataEngine, types.SettingNameV2DataEngine:
+			if err := sc.updateDataEngine(settingName); err != nil {
+				return errors.Wrapf(err, "failed to apply %v setting to Longhorn instance managers when there are attached volumes. "+
+					"It will be eventually applied", settingName)
+			}
+		case types.SettingNameGuaranteedInstanceManagerCPU, types.SettingNameV2DataEngineGuaranteedInstanceManagerCPU:
+			dataEngine := longhorn.DataEngineTypeV1
+			if settingName == types.SettingNameV2DataEngineGuaranteedInstanceManagerCPU {
+				dataEngine = longhorn.DataEngineTypeV2
+			}
+
+			detached, _, err := sc.ds.AreAllVolumesDetached(dataEngine)
+			if err != nil {
+				return errors.Wrapf(err, "failed to check volume detachment for %v setting update", settingName)
+			}
+
+			if !detached {
+				return &types.ErrorInvalidState{Reason: fmt.Sprintf("failed to apply %v setting to Longhorn components when there are attached volumes. It will be eventually applied", settingName)}
+			}
```

Comment:
In general, LGTM.
Jus one question: when to use `return errors.Wrapf`  or `return &types.ErrorInvalidState`

Reply:
When returning the error to `syncSetting` and `handleErr` will receive and handle the error, and `errors.Wrapf` will be `error` log level and `types.ErrorInvalidState` will be `trace` log level. That's the method we prevent logs from flooding.
Any concerns?